vm/actor/src/builtin/power/state.rs

```diff
@@ -1,205 +1,221 @@
-    ) -> Result<(), String> {
-        let mut claim = self
-            .get_claim(store, miner)?
-            .ok_or(format!("no claim for actor {}", miner))?;
+    ) -> Result<(), Box<dyn StdError>> {
+        let old_claim = get_claim(claims, miner)?
+            .ok_or_else(|| actor_error!(ErrNotFound; "no claim for actor {}", miner))?;
 
-        let old_nominal_power = claim.quality_adj_power.clone();
+        self.total_qa_bytes_committed += qa_power;
+        self.total_bytes_committed += power;
 
-        // update power
-        claim.raw_byte_power += power;
-        claim.quality_adj_power += qa_power;
-
-        let new_nominal_power = &claim.quality_adj_power;
+        let new_claim = Claim {
+            raw_byte_power: old_claim.raw_byte_power.clone() + power,
+            quality_adj_power: old_claim.quality_adj_power.clone() + qa_power,
+        };
 
         let min_power_ref: &StoragePower = &*CONSENSUS_MINER_MIN_POWER;
-        let prev_below: bool = &old_nominal_power < min_power_ref;
-        let still_below: bool = new_nominal_power < min_power_ref;
+        let prev_below: bool = &old_claim.quality_adj_power < min_power_ref;
+        let still_below: bool = &new_claim.quality_adj_power < min_power_ref;
 
         if prev_below && !still_below {
             // Just passed min miner size
-            self.num_miners_meeting_min_power += 1;
-            self.total_quality_adj_power += new_nominal_power;
-            self.total_raw_byte_power += &claim.raw_byte_power;
+            self.miner_above_min_power_count += 1;
+            self.total_quality_adj_power += &new_claim.quality_adj_power;
+            self.total_raw_byte_power += &new_claim.raw_byte_power;
         } else if !prev_below && still_below {
             // just went below min miner size
-            self.num_miners_meeting_min_power -= 1;
+            self.miner_above_min_power_count -= 1;
             self.total_quality_adj_power = self
                 .total_quality_adj_power
-                .checked_sub(&old_nominal_power)
-                .ok_or("Negative nominal power")?;
+                .checked_sub(&old_claim.quality_adj_power)
+                .expect("Negative nominal power");
             self.total_raw_byte_power = self
                 .total_raw_byte_power
-                .checked_sub(&claim.raw_byte_power)
-                .ok_or("Negative raw byte power")?;
+                .checked_sub(&old_claim.raw_byte_power)
+                .expect("Negative raw byte power");
         } else if !prev_below && !still_below {
             // Was above the threshold, still above
             self.total_quality_adj_power += qa_power;
             self.total_raw_byte_power += power;
         }
 
-        if self.num_miners_meeting_min_power < 0 {
-            return Err(format!(
-                "negative number of miners: {}",
-                self.num_miners_meeting_min_power
-            ));
-        }
+        assert!(
```

Comment:
I think there might be an assert missing, go code has these lines : 

AssertMsg(newClaim.RawBytePower.GreaterThanEqual(big.Zero()), "negative claimed raw byte power: %v", newClaim.RawBytePower)

AssertMsg(newClaim.QualityAdjPower.GreaterThanEqual(big.Zero()), "negative claimed quality adjusted power: %v", newClaim.QualityAdjPower)